File: AdventDotNet/Day5.cs
namespace AdventDotNet;

public static class Day5
{
    public static string SolvePart1(string[] rows)
    {
        var intervals = rows
            .Where(row => row.Contains('-'))
            .Select(row =>
            {
                var parts = row.Split('-');
                return (start: long.Parse(parts[0]), end: long.Parse(parts[1]));
            })
            .ToArray();
        var ingredients = rows
            .Where(row => !row.Contains('-') && !string.IsNullOrEmpty(row))
            .Select(long.Parse)
            .ToArray();

        var sum = ingredients.Count(i => intervals.Any(interval => interval.start <= i && i <= interval.end));
        return $"{sum}";
    }
    
    public static string SolvePart2(string[] rows)
    {
        var intervals = rows
            .Where(row => row.Contains('-'))
            .Select(row =>
            {
                var parts = row.Split('-');
                return (start: long.Parse(parts[0]), end: long.Parse(parts[1]));
            })
            .OrderBy(interval => interval.start)
            .ToList()
            .Merge();

        var sum = intervals.Sum(interval => interval.end - interval.start + 1);
        return $"{sum}";
    }

    private static List<(long start, long end)> Merge(this List<(long start, long end)> intervals)
    {
        var result = new List<(long start, long end)>();
        var candidate = intervals[0];

        for (var i = 1; i < intervals.Count; i++) {
            if (intervals[i].start <= candidate.end)
            {
                if (intervals[i].end > candidate.end)
                {
                    candidate.end = intervals[i].end;
                }
            }
            else
            {
                result.Add(candidate);
                candidate = intervals[i];
            }
        }

        result.Add(candidate);
        return result;
    }
}

File: src/day5.ts
export function solvePart1(rows: string[]): string {
    let intervals: Interval[] = rows
        .filter(row => row.includes("-"))
        .map(i => {
            let parts = i.split("-");
            return {
                start: BigInt(parts[0]!),
                end: BigInt(parts[1]!)
            };
        });
    let ingredients = rows
        .filter(row => !row.includes("-") && row.length > 0)
        .map(i => BigInt(i));

    let sum = ingredients.filter(i => intervals.some(interval => interval.start <= i && i <= interval.end))
        .length;

    return `${sum}`;
}

export function solvePart2(rows: string[]): string {
    let sum = 0n;

    let intervals: Interval[] = rows
        .filter(row => row.includes("-"))
        .map(i => {
            let parts = i.split("-");
            return {
                start: BigInt(parts[0]!),
                end: BigInt(parts[1]!)
            };
        })
        .sort((a, b) => {
            let diff = a.start - b.start;
            return diff < 0 ? -1 : (diff > 0 ? 1 : 0);
        });
    intervals = merge(intervals);

    intervals.forEach(interval => {
        sum += interval.end - interval.start + 1n;
    });

    return `${sum}`;
}

function merge(intervals: Interval[]) {
    let result: Interval[] = [];
    let candidate = intervals[0]!;

    for (let i = 1; i < intervals.length; i++) {
        if (intervals[i]!.start <= candidate.end) {
            if (intervals[i]!.end > candidate.end) {
                candidate.end = intervals[i]!.end;
            }
        } else {
            result.push(candidate);
            candidate = intervals[i]!;
        }
    }

    result.push(candidate);
    return result;
}

interface Interval {
    start: bigint,
    end: bigint
}